# Review

## Basic variable types

In [1]:
a = 1
b = [2, 3, 4, 5]
c = (6, 7)
d = {"a": 1, "b":2}
e = False
f = "Hello world"

In [2]:
len(b)

4

In [3]:
len(f)

11

In [4]:
len(d)

2

In [5]:
f[0]

'H'

In [6]:
f[3:5]

'lo'

In [7]:
f[-1]

'd'

In [8]:
f[:-2]

'Hello wor'

In [9]:
f[2:]

'llo world'

# Conditional logic

In [10]:
if a > 1:
    print("a is greater than 1")
elif a == 1:
    print("a equals 1")
else:
    print("a is less than 1")

a equals 1


## Looping

In [11]:
for letter in f:
    print("the letter is", letter)

the letter is H
the letter is e
the letter is l
the letter is l
the letter is o
the letter is  
the letter is w
the letter is o
the letter is r
the letter is l
the letter is d


In [12]:
for i in range(0,5):
    print(i)

0
1
2
3
4


In [13]:
for key,value in d.items():
    print("k:", key, "v:", value)

k: a v: 1
k: b v: 2


In [14]:
count = 0

while count < 5:
    print(count)
    count = count+1

0
1
2
3
4


In [15]:
for letter in f:
    if letter == "l":
        continue
    elif letter == "w":
        break
    print("the letter is", letter)


the letter is H
the letter is e
the letter is o
the letter is  


## Functions

In [16]:
def myfunc(a=1, b=2):
    print("a is", a, "b is", b)
    return a*b

In [17]:
myfunc(3,4)

a is 3 b is 4


12

## Comprehensions

In [18]:
[i**2 for i in range(0,5)]

[0, 1, 4, 9, 16]

In [19]:
tuple(i**2 for i in range(0,5))

(0, 1, 4, 9, 16)

In [20]:
{a:b for a,b in enumerate(f)}

{0: 'H',
 1: 'e',
 2: 'l',
 3: 'l',
 4: 'o',
 5: ' ',
 6: 'w',
 7: 'o',
 8: 'r',
 9: 'l',
 10: 'd'}

# More on strings

Strings can be delimited by single quotes ('...') or double quotes ("...") --- it doesn't make a difference.


In [21]:
print("I'm fine")

I'm fine


If you are using single quotes, and you want to include an apostrophe, you can "escape" it by using a backslash (\)

In [22]:
print('I\'m fine')

I'm fine


There are a range of backslash escape characters:

| Code | Result         |
|------|----------------|
| \\'  | Single quote   |
| \\"  | Double quote   |
| \\   | Backslash      |
| \n   | New line       |
| \r   | Carriage return|
| \t   | Tab            |
| \f   | Form feed      |
| \ooo | Octal value    |
| \xhh | Hexadecimal value |

The one you are most likely to use is \\n for a newline.


In [23]:
print('This is one line\nThis is a new line')

This is one line
This is a new line


Another way to write multiline strings is with three single or double quotes

In [24]:
multistring = """This is one line
This is another line"""

In [25]:
multistring

'This is one line\nThis is another line'

In [26]:
print(multistring)

This is one line
This is another line


### Formatting

In [27]:
name = 'John'
age = 20000

"Hello I'm {}, my age is {}".format(name, age)

"Hello I'm John, my age is 20000"

In [28]:
"Hello I'm {0}, my age is {1}".format(name, age)


"Hello I'm John, my age is 20000"

In [29]:
name = 'Elizabeth'
f'Hello {name}!'

'Hello Elizabeth!'

In [30]:
a = 5
b = 10
f'Five plus ten is {a + b} and not {2 * (a + b)}.'

'Five plus ten is 15 and not 30.'

In [31]:
a = 10000000
f"{a:,}"

'10,000,000'

In [32]:
a = 3.1415926
f"{a:.2f}"

'3.14'

In [33]:
a = 0.816562
f"{a:.2%}"

'81.66%'

Number formatting table

|Number|Format|Output|description|
|------|------|------|-----------|
|3.1415926|{:.2f}|3.14|Format float 2 decimal places|
|3.1415926|{:+.2f}|+3.14|Format float 2 decimal places with sign|
|-1|{:+.2f}|-1.00|Format float 2 decimal places with sign|
|2.71828|{:.0f}|3|Format float with no decimal places|
|4|{:0>2d}|04|Pad number with zeros (left padding, width 2)|
|4|{:x<4d}|4xxx|Pad number with x’s (right padding, width 4)|
|10|{:x<4d}|10xx|Pad number with x’s (right padding, width 4)|
|1000000|{:,}|1,000,000|Number format with comma separator|
|0.35|{:.2%}|35.00%|Format percentage|
|1000000000|{:.2e}|1.00e+09|Exponent notation|
|11|{:11d}|11|Right-aligned (default, width 10)|
|11|{:<11d}|11|Left-aligned (width 10)|
|11|{:^11d}|11|Center aligned (width 10)|

See https://www.w3schools.com/python/ref_string_format.asp for more.

# Another while loop example: Calculating square roots

In [34]:
import math

M = 4

diff = math.inf

tol = 1e-5

x = 3

print("{:^20} {:^20}".format("x_new", "diff"))
while diff > tol:
    x_new = (x+M/x)/2
    diff = abs(x - x_new)
    print(f'{x_new:20.15f} {diff:20.15f}')
    x = x_new

       x_new                 diff        
   2.166666666666667    0.833333333333333
   2.006410256410256    0.160256410256410
   2.000010240026215    0.006400016384042
   2.000000000026215    0.000010240000000
   2.000000000000000    0.000000000026215


Let's put that in a function

In [35]:
def my_sqrt(M, guess=0, tol=1.0e-5, verbose = False):
    import math
    diff = math.inf
    if verbose:
        print(f'Calculating the square root of {M}')
    if guess==0:
        x = M/2
    else:
        x = guess
    if verbose:
        print("{:^20} {:^20}".format("x_new", "diff"))
    while diff > tol:
        x_new = (x+M/x)/2
        diff = abs(x - x_new)
        if verbose:
            print(f'{x_new:20.15f} {diff:20.15f}')
        x = x_new
    if verbose:
        print(f'The square root of {M} is {x}')
    return x

my_sqrt(6, verbose=True)
my_sqrt(60, verbose=True)

Calculating the square root of 6
       x_new                 diff        
   2.500000000000000    0.500000000000000
   2.450000000000000    0.050000000000000
   2.449489795918367    0.000510204081633
   2.449489742783179    0.000000053135189
The square root of 6 is 2.4494897427831788
Calculating the square root of 60
       x_new                 diff        
  16.000000000000000   14.000000000000000
   9.875000000000000    6.125000000000000
   7.975474683544304    1.899525316455696
   7.749268935860040    0.226205747684264
   7.745967396017448    0.003301539842592
   7.745966692414866    0.000000703602582
The square root of 60 is 7.745966692414866


7.745966692414866

# File input and output 

In [36]:
f = open("data/lorem.txt")

In [37]:
f.read()

'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Maecenas porttitor congue massa. \nFusce posuere, magna sed pulvinar ultricies, purus lectus malesuada libero, sit amet commodo magna eros quis urna. \nNunc viverra imperdiet enim. Fusce est. Vivamus a tellus. \nPellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. \nProin pharetra nonummy pede. \nMauris et orci. \nAenean nec lorem. \nIn porttitor. \nDonec laoreet nonummy augue. \nSuspendisse dui purus, scelerisque at, vulputate vitae, pretium mattis, nunc. \nMauris eget neque at sem venenatis eleifend.\n'

It's good practice to close a file when you are finished with it.  Otherwise Python may hold onto resources which it could otherwise release.

In [38]:
f.close()

We can check if a file is open or closed.

In [39]:
f.closed

True

The more common practice is to use "context managers" which automatically close a file, and won't try to read from a file if you fail to open it.

In [40]:
with open("data/lorem.txt") as f:
    txt = f.read()
    

In [41]:
txt

'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Maecenas porttitor congue massa. \nFusce posuere, magna sed pulvinar ultricies, purus lectus malesuada libero, sit amet commodo magna eros quis urna. \nNunc viverra imperdiet enim. Fusce est. Vivamus a tellus. \nPellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. \nProin pharetra nonummy pede. \nMauris et orci. \nAenean nec lorem. \nIn porttitor. \nDonec laoreet nonummy augue. \nSuspendisse dui purus, scelerisque at, vulputate vitae, pretium mattis, nunc. \nMauris eget neque at sem venenatis eleifend.\n'

In [42]:
f.closed

True

## Reading a file line by line

In [43]:
with open("data/lorem.txt") as f:
    for line in f:
        print(line)
    

Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Maecenas porttitor congue massa. 

Fusce posuere, magna sed pulvinar ultricies, purus lectus malesuada libero, sit amet commodo magna eros quis urna. 

Nunc viverra imperdiet enim. Fusce est. Vivamus a tellus. 

Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. 

Proin pharetra nonummy pede. 

Mauris et orci. 

Aenean nec lorem. 

In porttitor. 

Donec laoreet nonummy augue. 

Suspendisse dui purus, scelerisque at, vulputate vitae, pretium mattis, nunc. 

Mauris eget neque at sem venenatis eleifend.



In [44]:
with open("data/lorem.txt") as f:
    line = f.readline()
    print(f"The first line is: \n\t{line}")
    

The first line is: 
	Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Maecenas porttitor congue massa. 



We can also use the **.split()** method to split a string into individual words.

In [45]:
with open("data/lorem.txt") as f:
    line = f.readline()
    words = line.split()
    print(f"The first line is: \n\t{line}")
    print(f"The first word is: '{words[0]}'")
    

The first line is: 
	Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Maecenas porttitor congue massa. 

The first word is: 'Lorem'


## Writing to a file

So far we've opened the file for reading, which is the default.  We can pass a second 'mode' argument to the **open** function to specify what we want to do with the file.

The possibilities are:

|mode symbol| Meaning|
|-----------|:-------|
|r          |Read|
|w          |Write (this will erase any existing file with this name).|
|a          |Append (data will be written at the end of an existing file|
|r+         |Read and write|

In addition you can append a "b" to the mode to specify that it is a binary file rather than a text file.  If it's a text file (the default) line endings will be converted to newlines (\\n) --- Windows uses a carriage return followed a newline (\\r\\n) in its text files. Similarly when writing it will convert newlines to the platform-specific line ending.  Setting the mode to binary prevents this conversion from happening, so avoids corruption of binary files.

In [46]:
with open("newfile.txt", "w") as f:
    f.write("A first line\n")
    f.write("A second line\n")

## CSV files

There is a standard Python library for dealing with CSV files.

In [47]:
import csv

In [48]:
with open("data/random2.csv", "r", newline='') as f:
    reader = csv.reader(f)
    for row in reader:
        print(row)

['50.3517896', '123.6495128']
['50.45255418', '123.5986666']
['51.50553258', '124.9399658']
['42.20666562', '107.2815025']
['49.27482836', '121.4525277']
['56.55078855', '134.6298668']
['52.48216726', '129.7236145']
['44.89864176', '112.9609078']
['54.65476358', '132.4543008']
['58.89295514', '141.6328837']
['50.55067454', '125.0597947']
['59.66646465', '143.1044854']
['53.69697921', '130.5292128']
['49.95601359', '122.1277307']
['46.28958975', '115.9963129']
['39.43666433', '100.6770527']
['38.3942892', '99.28664503']
['51.38637471', '127.6244957']
['46.61693519', '115.618257']
['44.96374067', '113.6807404']
['46.25737425', '115.7281136']
